samples/go/ipfs-chat/daemon.go

```diff
@@ -5,49 +5,136 @@
 package main
 
 import (
-	"fmt"
+	"context"
+	"encoding/base64"
 	"log"
 	"os"
 	"time"
 
 	"github.com/attic-labs/noms/go/d"
 	"github.com/attic-labs/noms/go/datas"
+	"github.com/attic-labs/noms/go/hash"
 	"github.com/attic-labs/noms/go/ipfs"
+	"github.com/attic-labs/noms/go/merge"
 	"github.com/attic-labs/noms/go/spec"
+	"github.com/attic-labs/noms/go/types"
 	"github.com/attic-labs/noms/samples/go/ipfs-chat/dbg"
+	"github.com/ipfs/go-ipfs/core"
 )
 
-func runDaemon(topic string, interval time.Duration, networkDS, localDS string, netNodeIdx, localNodeIdx int) {
+func runDaemon(topic string, interval time.Duration, ipfsSpec string, nodeIdx int) {
 	dbg.SetLogger(log.New(os.Stdout, "", 0))
 
-	ipfs.NodeIndex = netNodeIdx
-	sourceSp, err := spec.ForDataset(networkDS)
+	sp, err := spec.ForDataset(ipfsSpec)
 	d.CheckErrorNoUsage(err)
-	source := sourceSp.GetDataset()
-	source, err = InitDatabase(source)
-	d.PanicIfError(err)
+	ds := sp.GetDataset()
 
-	ipfs.NodeIndex = localNodeIdx
-	destSp, err := spec.ForDataset(localDS)
-	d.CheckErrorNoUsage(err)
-	dest := destSp.GetDataset()
-	dest, err = InitDatabase(dest)
-	d.PanicIfError(err)
+	node, cs := reconfigureIPFSChunkStore(sp, topic, nodeIdx)
+	sourceDB := datas.NewDatabase(cs)
+	sourceDS := sourceDB.GetDataset(ds.ID())
 
-	ipfs.NodeIndex = -1
+	cs = ipfs.NewChunkStorePrimitive(sp.DatabaseName, true, node)
+	destDB := datas.NewDatabase(cs)
+	destDS := destDB.GetDataset(ds.ID())
+	destDS, err = InitDatabase(destDS)
+	d.PanicIfError(err)
 
-	fmt.Printf("Replicating %s to %s...\n", sourceSp.String(), destSp.String())
+	dbg.Debug("Storing locally to:", sp.String())
 
-	sub, err := ipfs.CurrentNode.Floodsub.Subscribe(topic)
-	d.PanicIfError(err)
-	go Replicate(sub, source, dest, func(ds datas.Dataset) {
-		dest = ds
+	go replicate(node, topic, sourceDS, destDS, func(ds1 datas.Dataset) {
+		destDS = ds1
 	})
 
 	for {
-		s := dest.HeadRef().TargetHash().String()
-		fmt.Println("publishing: " + s)
-		Publish(sub, topic, s)
+		Publish(node, topic, destDS.HeadRef().TargetHash())
 		time.Sleep(interval)
 	}
 }
+
+// replicate continually listens for commit hashes published by ipfs-chat nodes,
+// ensures that all nodes are replicated locally, and merges new data into it's
+// dataset when necessary.
+func replicate(node *core.IpfsNode, topic string, sourceDS, destDS datas.Dataset, didChange func(ds datas.Dataset)) {
```

Comment:
What's the difference between this and `MergeMessages`? It looks like one calls `Pull` and the other doesn't. If that's it, or if they are very close they should probably share code so that they don't drift apart.

Reply:
I agree, they're pretty similar but at this point I think it's helpful to be able to work with them independently until we find the magic sauce. Then we can merge them again.